This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [1]:
# get api key from your .env file
import os
from dotenv import load_dotenv

load_dotenv()
API_KEY = os.getenv('NASDAQ_API_KEY')

Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [16]:
# First, import the relevant modules
import requests
import os
from dotenv import load_dotenv

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [17]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
response=requests.get("https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X",params={'api_key': API_KEY,
    'start_date': '2020-01-01',
    'end_date': '2020-01-01'})
json_data=response.json()

In [18]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
print(json_data)

{'dataset': {'id': 10095370, 'dataset_code': 'AFX_X', 'database_code': 'FSE', 'name': 'Carl Zeiss Meditec (AFX_X)', 'description': 'Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704', 'refreshed_at': '2020-12-01T14:48:09.907Z', 'newest_available_date': '2020-12-01', 'oldest_available_date': '2000-06-07', 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'frequency': 'daily', 'type': 'Time Series', 'premium': False, 'limit': None, 'transform': None, 'column_index': None, 'start_date': '2020-01-01', 'end_date': '2020-01-01', 'data': [], 'collapse': None, 'order': None, 'database_id': 6129}}


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [19]:
# Collect data for the whole year 2017
r=requests.get("https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X",params={'api_key': API_KEY,
    'start_date': '2017-01-01',
    'end_date': '2017-12-31'})
d=r.json()
N_data=d['dataset']

In [20]:
data=list(N_data['data'])

In [21]:
# Extract opening prices and find the highest and lowest opening prices
opening_price=[row[1] for row in data ]

In [22]:
highest_opening_price=max([d for d in opening_price if d is not None])
highest_opening_price_index=opening_price.index(highest_opening_price)
date_highest_opening_price=data[highest_opening_price_index][0]
print(f"The highest opening price is found on the date of {date_highest_opening_price} and its value is {highest_opening_price}")

The highest opening price is found on the date of 2017-12-14 and its value is 53.11


In [23]:
lowest_opening_price=min([d for d in opening_price if d is not None])
lowest_opening_price_index=opening_price.index(lowest_opening_price)
date_lowest_opening_price=data[lowest_opening_price_index][0]
print(f"The lowest opening price is found on the date of {date_lowest_opening_price} and its value is {lowest_opening_price}")

The lowest opening price is found on the date of 2017-01-24 and its value is 34.0


In [24]:
# Calculate the largest change in any one day based on High and Low price
high_price=[row[2] for row in data]
low_price=[row[3] for row in data]
diff=[high - low for high, low in zip(high_price, low_price)]
largest_day_change_index=diff.index(max(diff))
date_largest_day_change=data[largest_day_change_index][0]
print(f"The largest change in any one day is found on the date of {date_largest_day_change} and its value is {max(diff)}")

The largest change in any one day is found on the date of 2017-05-11 and its value is 2.8100000000000023


In [25]:
# Calculate the largest change between any two days based on Closing Price
close_price=[row[4] for row in data]

In [26]:
change=[0]*254
for n in range(len(close_price)-1):
    change[n]=close_price[n+1]-close_price[n]

In [27]:
largest_change_two_days=max([abs(a) for a in change])
largest_change_two_days_index=change.index(largest_change_two_days)
date_largest_change_two_days=data[largest_change_two_days_index][0]
print(f"The largest change between two days is found between {data[largest_change_two_days_index+1][0]} and {date_largest_change_two_days} and its value is {largest_change_two_days}")

The largest change between two days is found between 2017-08-08 and 2017-08-09 and its value is 2.559999999999995


In [28]:
# Calculate the average daily trading volume
daily_trading_volume=[row[6] for row in data]
average_daily_trading=sum(daily_trading_volume)/len(daily_trading_volume)
print(f"The average daily trading volume is: {average_daily_trading}")

The average daily trading volume is: 89124.33725490196


In [29]:
#function to calculate the median
def calculate_median(data):
    sorted_data = sorted(data)
    n = len(sorted_data)
    if n % 2 == 0:
        return (sorted_data[n // 2 - 1] + sorted_data[n // 2]) / 2
    else:
        return sorted_data[n // 2]

# Calculate and print the median trading volume
median_volume = calculate_median(daily_trading_volume)
print(f"The median trading Volume: {median_volume}")


The median trading Volume: 76286.0


Summary: The opening prices for the stock revealed a highest value of 53.11 on 2017-12-14 and a lowest value of 34.0 on 2017-01-24. Daily price changes were examined, highlighting the largest change in any one day on 2017-05-11, with a value of 2.81, and the largest change between any two consecutive days, occurring between 2017-08-08 and 2017-08-09, with a value of 2.56.
Additionally, the average daily trading volume for the year was calculated to be approximately 89,124.34. Furthermore, the median trading volume, determined to be 76,286.0.